clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8714,6 +8714,21 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
     }
   }
 
+  // zero sized static arrays are not allowed in HIP device functions
+  if (LangOpts.CUDAIsDevice && LangOpts.HIP) {
+    if (FunctionDecl *FD = getCurFunctionDecl();
+        FD &&
+        (FD->hasAttr<CUDADeviceAttr>() || FD->hasAttr<CUDAGlobalAttr>())) {
```

Comment:
Is there not some common helper for "will the current context be built for device"?  This seems like it should be common enough to show up in multiple places.

Reply:
Theres one for CUDA, but I didn't see one for LangOpts.HIP.

Reply:
I'd prefer some reuse, even if you need some additional checks.

Reply:
The reuse of that DeclAttrsMatchCUDAMode with some rechecks doesn't make sense because that function returns true if decl is not defined. So, inturn I have to check if its inside a functionDecl again else it will wrongly diagnose host global zero array as not permitted.